python/GafferImageUI/CatalogueUI.py

```diff
@@ -371,28 +375,55 @@ def __addClicked( self, *unused ) :
 
 	def __removeClicked( self, *unused ) :
 
-		index = self.__indexFromSelection()
+		indices = self.__indicesFromSelection()
 
 		# If the user repeatedly clicks the delete button, we might end up in a
 		# state, where selection hasn't been restored yet. In that case we
 		# can't delete anything and will ignore the request.
-		if index is None :
+		if not indices :
 			return
 
 		with Gaffer.UndoScope( self.getPlug().ancestor( Gaffer.ScriptNode ) ) :
-			self.__images().removeChild( self.__images()[index] )
-			self.getPlug().setValue( max( 0, index - 1 ) )
+			for i, index in enumerate( reversed( sorted( indices ) ) ) :
+				self.__images().removeChild( self.__images()[index] )
+			self.getPlug().setValue( max( 0, index-1 ) )
 
 	def __extractClicked( self, *unused ) :
 
-		index = self.__indexFromSelection()
-		image = self.__images()[index]
+		indices = self.__indicesFromSelection()
+		for index in indices :
+			image = self.__images()[index]
+
+			plugParent = self.__catalogue()
+			outPlug = plugParent["out"]
+
+			# It's possible that the Catalogue was wrapped by another node
+			if self.getPlug().parent() is not self.__catalogue():
```

Comment:
This does work, but the logic was a bit hard to follow. I think the parent for the CatalogueSelect is always `self.getPlug().node().parent()`, so doesn't need to be computed from the walk. And I think the walk can be achieved in reverse as a one liner : `outPlug = next( p for p in node.children( GafferImage.ImagePlug ) if catalogue.isAncestorOf( p.source() ) )`. And these are both constant so can be lifted out of the loop, giving something much more straightforward :

```
node = self.getPlug().node()
catalogue = self.__catalogue()
outPlug = next( p for p in node.children( GafferImage.ImagePlug ) if catalogue.isAncestorOf( p.source() ) )

for index in self.__indicesFromSelection() :
	image = self.__images()[index]

	extractNode = GafferImage.CatalogueSelect()
	extractNode["in"].setInput( outPlug )
	extractNode["imageName"].setValue( image.getName() )

	node.parent().addChild( extractNode )
```